### ISSUE TITLE
Clean up the background map list

### ISSUE THREAD

**Issue by McPringle** [zdila/gpx-animator#61]
Some maps did not work anymore. It is annoying to go through the maps one by one to find a working one. Deprecated, not working maps should be removed.

**Comment by McPringle** [comment]
Maps, which cover only a national or regional part, should be named accordingly.

**Comment by n76** [comment]
I am currently using v1.3.1 and many backgrounds are not working. I don't know if it is pertinent, but if I run gpx-animator from the command line on a background that is failing it will give me an error message of the form:

`error reading tile https://b.tile.openstreetmap.org/15/5674/13064.png`

If I try to download the indicated map tile manually using wget, it will download fine:

```
$ wget https://b.tile.openstreetmap.org/15/5674/13064.png
--2019-07-10 20:25:50--  https://b.tile.openstreetmap.org/15/5674/13064.png
Resolving b.tile.openstreetmap.org (b.tile.openstreetmap.org)... 140.211.167.105, 204.16.246.252
Connecting to b.tile.openstreetmap.org (b.tile.openstreetmap.org)|140.211.167.105|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 9035 (8.8K) [image/png]
Saving to: ‘13064.png’

13064.png                                    100%[===========================================================================================>]   8.82K  8.82KB/s    in 1.0s    

2019-07-10 20:25:51 (8.82 KB/s) - ‘13064.png’ saved [9035/9035]
```

And I can load the tile using my web browser too.

So the link is good. Not sure if this helps or not but it seemed odd to me that good link should fail.

**Comment by n76** [comment]
Interesting about not handling things like redirects. Not sure if it is the case I posted above as I don't see that wget got a redirect, but it would make things more robust.

I am a novice at Java only using it within an Android app. But if I can help, I'd like to. In the case of this project, I don't see any equivalent of a make or build file as I'd expect for a C, C++ or Android APK build environment. How does one build this into a JAR file? -Thanks!

**Comment by McPringle** [comment]
Hi @n76, handling of redirects has pros and cons. First, technically, GPX Animator uses a JDK method to load resources via an URL which does not handle redirects. This part has to be rewritten to use e.g. an Apache Commons library to fetch the tile resources. That would fix this problem but create a new one:

If an URL is modified and a redirect points to the new location and GPX Animator follows this redirect, nobody would recognize that GPX Animator still tries the old URLs first. If, for example, 50 tiles are needed this usually results in 50 HTTP requests. If there is a redirect, this will double the requests and then 100 requests are needed. That makes loading tiles slow and inefficient.

A better solution would be e.g. follow the redirect and load the tiles but issue a descriptive warning to the user with the old and new URLs, so the gets aware of this problem and can solve it (if he entered the URL directly) or report it (to fix the URL in GPX Animator itself). But that is, of course, more work.

A very luxury solution would be to provide a "Report" functionality in the warning dialog which either creates a GitHub issue automatically (needs GitHub authentication) or just sends an email to a collector email address.

**Comment by McPringle** [comment]
About the build: This project contains a `pom.xml` which is the build file for Maven. I will modify this later to add a Maven wrapper. With this wrapper, everyone can execute the build without the need to install Maven.

To provide some more information now, I added the build instructions to the `README.md` file.

**Comment by McPringle** [comment]
> 50 tiles would be needed only for video with big dimensions. Videos of regular dimensions uses only like 20 tiles and the redirects should not be the problem.

For sure, tile caching will reduce the problem of redirects a bit. A few weeks ago I ran into "too many connections from your IP address" problems on OSM with my Barcelona video. That's why I am a little sensitive about redirects. We can go and solve this problem by step. Nothing needs to be perfect at the first shot... :-)

**Comment by n76** [comment]
> About the build: This project contains a `pom.xml` which is the build file for Maven. 

Thanks for that. I was able to figure out that pom.xml files were for Maven and after installing Maven (and some other items) am able to build a working version from the source repository.

Having a "how to build" section in the README.md file is great.

**Comment by n76** [comment]
re: caching map tiles. I too ran into it using OSM tiles when playing around with various options to see what the results were, so I think it would be nice to implement.  I don't think that would be too hard to implement, the difficult part being deciding on behavior. May I suggest:

- Cache to be some place like ~/.gpx_animator/cache (what would be the DOS/Windows equivalent?)
- That the cache files be named based on a hash of the full tile URL, that way the cache could hold tiles from several different TMS providers.
- A new CLI/UI option for specifying maximum age of cache (older files removed).

Maybe create a new Cache.java to implement the cache and then have Map.java use it to access the map tiles. OTOH, cache logic is pretty simple so maybe it should just be done in Map.java. Anyway I could probably implement that tomorrow . . . assuming this middle of the night bout of insomnia abates and I get some sleep :)

**Comment by zdila** [comment]
Caching is never simple (eg. stale cache...). The best would be to use existing HTTP library that leverages caching.

**Comment by zdila** [comment]
Still there are pretty much HTTP headers to consider: https://www.keycdn.com/blog/http-cache-headers

**Comment by McPringle** [comment]
There could be a Unit Test which executes a request to each map in the `maps.xml` file and check the HTTP response. If the response is different from `200 OK`, the test should fail. The implementation for such a test should not be too complicated and the test should be able to identify shut down or moved map resources.

**Comment by McPringle** [comment]
I'm going to implement a very simple solution: Every map entry needs to have a validation URL defined. This will be checked during the test phase of the build. If the request to the URL fails, the build will fail.

**Comment by McPringle** [comment]
This is how it works now: There is a new file 'maps-tests.xml' in the folder 'src/test/resources' which contains validation URLs for the maps. There can be multiple URLs per map, there is no limitation. In the _test_ phase of the build, there will be one request for each URL. All requests have to return a tile (an image). If not, the test for this map fails. The whole test fails if at least one of the configured validation URL fails. The build will stop. Maps without a validation URL will _not_ make the build fail.

On the console, there will be a detailed report of which maps do not have tests, which maps failed the test, and which maps successfully passed the test. This is the first step in cleaning up the list of maps and removing broken ones.

Next step: Add validation URLs for all maps (as in the example of _OpenStreetMap (German Style)_) and remove the maps, which do no longer work.

**Comment by McPringle** [comment]
Btw: There are **333** maps in GPX Animator! :astonished:

**Comment by McPringle** [comment]
**Update**
- I created tests for 51 maps.
- I had to remove 8 maps because they are no longer providing their service (or are now behind a paywall).
- I was able to fix 20 broken maps, which are now working again.
- There are still 282 maps without a test.

It is very time-consuming creating tests for regional maps that cover only a limited area because I can't reuse my tile information and have to check for local tiles. I'll implement a "Report Map" feature in GPX Animator which allows the user to report a map as broken or as working.
